### PR TITLE
Upgrade distributed test to g4dn instances (T4 GPUs)

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137763
* #135273
* __->__ #137161

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Review by seemethere — APPROVED** [review]
LGTM, nit: since these nodes are 2x as expensive it might be good to move the experimental split build duplicate tests to periodic.yml.

WDYT @PaliC

**Comment by PaliC** [comment]
@pytorchbot revert -m "broken tests on trunk" -c "nosignal"
check out https://hud.pytorch.org/pytorch/pytorch/commit/b6a64dce072240c0b06d2fb03ac81b3ed1b73d92

File: test/run_test.py
Comment (kwen2501) [inline test/run_test.py:312]: Cc @Aidyn-A @eqy 
I was trying to upgrade the CI infra from 2x M60 to 4x T4.
During the upgrade, some UCC tests seems unstable. I did not investigate much cuz there are other failures to unblock too.
So I just temporarily disabled testing `test/distributed/test_distributed_spawn.py` with UCC backend.
Longer term, I think we can discuss more about PyTorch's support for UCC backend.

File: torch/testing/_internal/distributed/distributed_test.py
Comment (eqy) [inline torch/testing/_internal/distributed/distributed_test.py:10003]: This looks suspicious, is there a reason this should be hardcoded?

Comment (kwen2501) [reply torch/testing/_internal/distributed/distributed_test.py:10003]: I thought here we should do / 2 because the forward + backward is run twice, which is unrelated to world_size.

Setting to / 2 and run locally the test passed.

Comment (eqy) [reply torch/testing/_internal/distributed/distributed_test.py:10003]: I see, makes sense, maybe it's clearer to make the iteration count a constant and share it with the previous loop?

**Review by eqy — APPROVED** [review]

**Comment by eqy** [comment]
CC @nWEIdia @ptrblck  as it looks like we may be moving off of M60 runners!

**Comment by kwen2501** [comment]
@eqy I remember there was a succinct CUDA Graph implementation using thread local mode, which M60 didn't like. Maybe we can bring that back after the GPU upgrade? Not urgent though.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change, but it seems another failure showing up after the upgrade' -c weird

distributed/_tools/test_runtime_estimator.py::TestRuntimeEstimator::test_transformer_runtime [GH job link](https://github.com/pytorch/pytorch/actions/runs/11359220817/job/31596426837) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/f43c4d28b8f955fe1f2b80f193815edadc95507b)

The test is currently disabled on ROCm https://github.com/pytorch/pytorch/issues/134824 (not linux though), so maybe there are some clues there.

**Comment by kwen2501** [comment]
@huydhn Oh I thought the test had been disabled. 
Screenshot 2024-10-16 at 9 18 40 AM
In any case, let me disable the check in the test code.

**Comment by huydhn** [comment]
Let me add `ci-no-td` onto this PR to make sure that all tests are running

**Review by yf225 — CHANGES_REQUESTED** [review]
Would it be okay to not land this PR yet, and instead fix the tests that need to be disabled? Those tests are guarding all of the Compiled DDP / Compiled FSDP functionalities, and they are very easy to regress if we don't have unit tests for them.

I am happy to help look into these disabled tests. What would be the best way to access a g4dn machine? I can go with my personal AWS account to debug them.

cc. @fegin

**Comment by kwen2501** [comment]
Thanks @yf225 for the heads up.
The failure of `TestFullyShardCompile` class does not need g4dn machines to repro. 
I can repro them on my 4x H100 machine locally.
`python test/distributed/_composable/fsdp/test_fully_shard_compile.py`

**Comment by yf225** [comment]
Discussed offline:

For Compiled FSDP errors: https://github.com/pytorch/pytorch/issues/138147#issuecomment-2418306639
- It's likely only the bf16 triton codegen related error - I will change the tests to not use bf16 when <sm80

For Compiled DDP errors: I can find the error trace in https://ossci-raw-job-status.s3.amazonaws.com/log/31640268847:
- test_ddp_tp: we can skip it for now
- test_compile_bf16: bf16-related error
- test_compile_backward_only: can be fixed by @fegin's upcoming PR

**Review by yf225 — APPROVED** [review]
Thanks a lot!

**Pull request closed without merging**
